distributed/utils_comm.py

```diff
@@ -77,36 +116,46 @@ async def gather_from_workers(
                         address,
                         who=who,
                         serializers=serializers,
-                        max_connections=False,
                     ),
                     operation="get_data_from_worker",
                 ),
                 name=f"get-data-from-{address}",
             )
             for address, keys in d.items()
         }
-        response: dict[str, object] = {}
-        for worker, c in coroutines.items():
+        for address, task in tasks.items():
             try:
-                r = await c
-            except OSError:
-                missing_workers.add(worker)
-            except ValueError as e:
-                logger.info(
-                    "Got an unexpected error while collecting from workers: %s", e
+                r = await task
+            except (OSError, asyncio.CancelledError, asyncio.TimeoutError):
+                # Note: CancelledError and asyncio.TimeoutError are rare conditions
+                # that can be raised by the network stack.
+                # See https://github.com/dask/distributed/issues/8006
+                missing_workers.add(address)
+            except Exception:
+                # For example, deserialization error
+                logger.exception(
+                    "Unexpected error while collecting tasks %s from %s",
+                    d[address],
+                    address,
                 )
-                missing_workers.add(worker)
+                for key in d[address]:
+                    missing_keys.add(key)
+                    del to_gather[key]
```

Comment:
Potentially only one of these keys caused the exception. Would it make sense to attempt gathering them individually to keep the # missing keys to a minimum? 

We may also consider returning these as erring keys, not missing. However, this is out of scope.

Reply:
Changed code to mark keys as erring, meaning that gather() won't try again.
However, note that there is no transition memory->error in the scheduler; this is the same problem as with gather_dep (https://github.com/dask/distributed/issues/6705).

Yes, we *could* try a fetch the keys one by one, but it feels like a substantial complication and ultimately overkill for what, *in theory*, should be a rare-ish problem easily weeded out during development?